[Core] Split up math_2d.h

aaronfranke opened the pull request:
I've split up `math_2d` into separate files for `vector2`, `rect2`, and `transform_2d`, and changed the relevant includes. I also renamed `Point2i` to `Vector2i`. Including `transform_2d.h` will include all the rest of the files, as it includes `rect2.h` which includes `vector2.h`.

Also, I've tweaked some grammar, and I've moved the math defs to `math_defs.h`. 

These changes improve the readability of the math code. CC @reduz @akien-mga See: https://github.com/godotengine/godot/issues/5691

groud commented:
I'm not against it, but why moving Vector2 and not Rect, Transform2D, Point2i (which should be renamed Vector2i someday for consistency) or Rect2i ?

This is quite inconsistent. Either we put all classes in the same file, either we split them all.

aaronfranke commented:
Rect2 and Transform2D refer to each other inside of `math_2d`, so I'm not sure of the best way to go about splitting them. I would not rename Point2i to Vector2i because does not behave like a vector. 

My idea is to eventually change all instances of `Point2` into `Vector2` and eventually change `Point2i` to `Point2`. I think it would be more useful to have these terms refer to different things, instead of to the same thing. And `Point2` vs `Point2i` is only one letter difference, it would be better for readability too if we had `Point` exclusively refer to integers and `Vector` exclusively refer to floats. But this is a discussion for another thread. The first step IMO is to organize existing stuff like `Vector2`.

groud commented:
Honestly I don't think that, as is, it improves readability. 
The current code organisation is fine to me. It's simple to include it as a single file, I'm not sure it's a gain splitting them, especially because it makes sinces more inconsistent in my opinion. Since, from my experience, it's rare having to use Vector2 without the remaining of the 2D tools (at least in the core/editor code)

aaronfranke commented:
What's inconsistent is the fact that 3D classes get their own files, but 2D is all in one.

There are plenty of times where Vector2 is used but not the rest of `math_2d`. For example, when working with screen positions, or when translating objects, or with texture UVs. Vector2 is used in 3D.

aaronfranke commented:
I found 7 other files where `Vector2` is used but the rest of `math_2d` is not, and I've made a fixup to my commit which changes the `#include` of these files to `"vector2.h"`.

I'm not sure if this helps with performance but it definitely is more organized to only include what's needed.

aaronfranke commented:
I made a minor fixup to move the [new C# constants from MathfEx](https://github.com/godotengine/godot/pull/18038) to Mathf since it's now in C++ too.

Chaosus commented:
In C++ but not in GDScript ? The purpose of MathfEx is to divide the GDScript / C#(not C++) open constant/methods and extensions from each other. And whats the Math_SQRT12 and Math_LN2 ?

Chaosus commented:
If you so want to add a new constant to engine than you should watch this example - https://github.com/godotengine/godot/pull/12627/files and do like this, and adding something new only to C++ does not make sense

Chaosus commented on thirdparty/misc/triangulator.h:
```diff
@@ -21,9 +21,9 @@
-#include "math_2d.h"
 #include "list.h"
 #include "set.h"
+#include "vector2.h"
```
You could include "vector2.h" as well as other divided classes into "math_2d.h".. and math_2d.h could be collection of "include". And no need to change this in such case..

Chaosus replied:
However this approach could improve readability.. its your choise, I just mentioned about this possibility

aaronfranke replied:
That is how it currently is, `math_2d.h` has an include for `vector2.h`. But if you're not using the rest of `math_2d.h` it would be better organized to only include what you need, yes?

aaronfranke commented:
Good point, I forgot for a bit what was the point of splitting (GDscript lacking it, not C++). I may attempt adding those to GDscript in another PR, but for now, I'll revert those Mathf/MathfEx changes.

The rest of this PR should be good, though.

reduz commented:
At this point I'm completely lost about what PR is about

aaronfranke commented:
@reduz It's primarily about moving Vector2 to its own file, for consistency with Vector3 etc. I made a few other minor changes as well so I just put them in this PR (rather than making two PRs).

If you'd like, I can split the Vector2 move and everything else into two commits on this PR.

aaronfranke committed:
[Core] Move Vector2 to its own file

Makes 2D math code easier to read and makes Vector2 consistent with Vector3. In the future, we may move other things out of math_2d as well.

aaronfranke committed:
[Core] Change math_2d includes to vector2 includes where relevant

aaronfranke committed:
[Core] Grammar, move defs to defs

Math constant definitions belong in `math_defs.h`. Also, the grammer in these files really needed some updating.

reduz commented:
I would do something like this:

1) kill math_2d
2) move Vector2, Vector2i to vector2.h
3) move Rect2 and Rect2i to rect2.h
4) move Transform2D to transform_2d,h

akien-mga commented:
5. move Point2 and Point2i to vector2.h too

aaronfranke commented:
@reduz Do you want me to do all that (killing math_2d entirely) in this pull request?

aaronfranke committed:
[Core] Move Rect2 and Transform2D to their own files

Math2D includes Transform2D, which includes Rect2, which includes Vector2.

aaronfranke committed:
[Core] Move Vector2i and Rect2i out of math_2d.h

aaronfranke committed:
[Core] Completely kill math_2d.h, change includes

aaronfranke commented:
@reduz I've done all of that now, `math_2d` is dead!
